### PR TITLE
Bugfixes for vulkan

### PR DESCRIPTION
**Pull request by sushantojal**
Vulkan bugfixes for runnning tests. Screenshots now work with vulkan, so they can be compared with golden master images.

Current status of framework tests with these changes:

Passed:
- BoundingVolumeTests
- CullTests
- RenderConfigTests

Partial (listing the failed sub-tests):
- PickerTests: canSendEventsToHitObjects

- RenderShaderTests:  meshFormatsWithShaderCombinationsTest

- RenderTests:	testTwoPostEffects
		        testBlendFunc

- SceneObjectTests: canDisplayCubes
		               canDisplayCylinders
		              canDisplayNonTextured
		              canDisplaySpheres

Crashes / Not Passsed:
- LightTests
- MiscTests
- ShadowTests
- StencilTests
- TextureTests
- TextureTransparencyTests
- VideoTests

### CODE REVIEW

**Comment by liaxim** [comment]
Do you see a lot of crashes? Can you share some stack traces? Thanks.

**Comment by sushantojal** [comment]
Light, shadows, video fail due to unsupported features. Stencil tests have a large difference from golden.
Crashes are there in the texture tests:

`signal 6 (SIGABRT), code -6 (SI_TKILL), fault addr --------
03-06 10:45:02.498 22876-22876/? A/DEBUG: Abort message: '/home/sushant.o/GVRF/GearVRf/GVRf/Framework/framework/src/main/jni/vulkan/vk_imagebase.cpp:152: int gvr::vkImageBase::updateMipVkImage(uint64_t, std::vector<void*>&, std::vector<gvr::ImageInfo>&, std::vector<VkBufferImageCopy>&, VkImageViewType, VkFormat, int, VkImageCreateFlags): assertion "formatProperties.optimalTilingFeatures & VK_FORMAT_FEATURE_BLIT_DST_BIT" failed'
03-06 10:45:02.498 22876-22876/? A/DEBUG:     r0 00000000  r1 0000592b  r2 00000006  r3 00000008
03-06 10:45:02.498 22876-22876/? A/DEBUG:     r4 cb5fb978  r5 00000006  r6 cb5fb920  r7 0000010c
03-06 10:45:02.498 22876-22876/? A/DEBUG:     r8 cb5f9820  r9 ca971500  sl 00000000  fp cb5f97a4
03-06 10:45:02.498 22876-22876/? A/DEBUG:     ip 0000000b  sp cb5f8b28  lr f02bbba7  pc f02be4a0  cpsr 200f0010
03-06 10:45:02.510 22880-22880/? E/Zygote: v2
03-06 10:45:02.511 22876-22876/? A/DEBUG: backtrace:
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #00 pc 0004b4a0  /system/lib/libc.so (tgkill+12)
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #01 pc 00048ba3  /system/lib/libc.so (pthread_kill+34)
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #02 pc 0001e1b9  /system/lib/libc.so (raise+10)
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #03 pc 000197a5  /system/lib/libc.so (__libc_android_abort+34)
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #04 pc 00017540  /system/lib/libc.so (abort+4)
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #05 pc 0001c153  /system/lib/libc.so (__libc_fatal+22)
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #06 pc 00019b87  /system/lib/libc.so (__assert2+18)
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #07 pc 006129c1  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr11vkImageBase16updateMipVkImageEyRSt6vectorIPvSaIS2_EERS1_INS_9ImageInfoESaIS6_EERS1_I17VkBufferImageCopySaISA_EE15VkImageViewType8VkFormatij+276)
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #08 pc 0060ba0d  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr13VkBitmapImage16updateFromMemoryEi+504)
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #09 pc 0060b7d9  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr13VkBitmapImage6updateEi+56)
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #10 pc 005a519b  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr5Image14checkForUpdateEi+78)
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #11 pc 0060b3b9  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr13VkBitmapImage7isReadyEv+16)
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #12 pc 00603491  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr7Texture7isReadyEv+36)
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #13 pc 006174d5  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr9VkTexture7isReadyEv+12)
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #14 pc 005de239  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr10ShaderData9updateGPUEPNS_8RendererEPNS_10RenderDataE+84)
03-06 10:45:02.511 22880-22880/? E/Zygote: accessInfo : 0
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #15 pc 005d4fbb  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr10RenderPass7isValidEPNS_8RendererERKNS_11RenderStateEPNS_10RenderDataE+98)
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #16 pc 005f3469  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr10RenderData7isValidEPNS_8RendererERKNS_11RenderStateE+140)
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #17 pc 005ac285  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr8Renderer13addRenderDataEPNS_10RenderDataERNS_11RenderStateERSt6vectorIS2_SaIS2_EE+28)
03-06 10:45:02.511 22876-22876/? A/DEBUG:     #18 pc 005ac333  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr8Renderer19occlusion_cull_initERNS_11RenderStateERSt6vectorIPNS_11SceneObjectESaIS5_EEPS3_IPNS_10RenderDataESaISA_EE+134)
03-06 10:45:02.512 22876-22876/? A/DEBUG:     #19 pc 005b241d  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr14VulkanRenderer14occlusion_cullERNS_11RenderStateERSt6vectorIPNS_11SceneObjectESaIS5_EEPS3_IPNS_10RenderDataESaISA_EE+32)
03-06 10:45:02.512 22876-22876/? A/DEBUG:     #20 pc 005ac1f9  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr8Renderer14cullFromCameraEPNS_5SceneEP8_jobjectPNS_6CameraEPNS_13ShaderManagerEPSt6vectorIPNS_10RenderDataESaISB_EEb+756)
03-06 10:45:02.512 22876-22876/? A/DEBUG:     #21 pc 005f5b4b  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr12RenderTarget14cullFromCameraEPNS_5SceneEP8_jobjectPNS_6CameraEPNS_8RendererEPNS_13ShaderManagerE+58)
03-06 10:45:02.512 22876-22876/? A/DEBUG:     #22 pc 005f7359  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (Java_org_gearvrf_NativeRenderTarget_cullFromCamera+124)
03-06 10:45:02.512 22876-22876/? A/DEBUG:     #23 pc 00631d89  /data/app/org.gearvrf.tester-1/oat/arm/base.odex (offset 0x5f9000)
`

**Comment by liaxim** [comment]
@apjagdale Please review

**Comment by roshanch** [comment]
code seems okay to me.

**Comment by apjagdale** [comment]
@liaxim I will review it today.

File: GVRf/Framework/framework/src/main/jni/view_manager_jni.cpp [hunk 1]
```diff
@@ -96,12 +99,26 @@ JNIEXPORT void JNICALL Java_org_gearvrf_GVRViewManager_readRenderResultNative(JN
...
     uint8_t *readback_buffer = (uint8_t*) env->GetDirectBufferAddress(jreadback_buffer);
     RenderTarget* renderTarget = reinterpret_cast<RenderTarget*>(jrenderTarget);
-    RenderTexture* renderTexture =    renderTarget->getTexture();
+    RenderTexture* renderTexture;
+
+    gRenderer = Renderer::getInstance();
+    if(gRenderer->isVulkanInstance())
+    {
+        uint8_t *data;
+        VkRenderTextureOffScreen* renderTexture = static_cast<VkRenderTextureOffScreen*>(static_cast<VkRenderTarget*>(renderTarget)->getTexture());
```
[the remark this answers is not in the document]
Comment (roshanch) [reply GVRf/Framework/framework/src/main/jni/view_manager_jni.cpp]: yes, he is going to change it.

[the remark this answers is not in the document]
Comment (sushantojal) [reply GVRf/Framework/framework/src/main/jni/view_manager_jni.cpp]: Should be ok now.

File: GVRf/Framework/framework/src/main/jni/vulkan/vk_render_target.h [hunk 2]
```diff
@@ -31,6 +31,8 @@ class VkRenderTarget: public  RenderTarget
     explicit  VkRenderTarget(){}
     virtual ~VkRenderTarget(){}
     virtual void    beginRendering(Renderer* renderer);
+
+    RenderTexture* getTexture();
```
Comment (apjagdale) [inline GVRf/Framework/framework/src/main/jni/vulkan/vk_render_target.h]: make it VkRenderTexture

**Comment by apjagdale** [comment]
Ran tests and post effect demo, both work.

**Comment by liaxim** [comment]
Looks good. Planning to merge on Monday.

**Pull request merged**